### PR TITLE
Stop training on Inf/NaN loss

### LINKED ISSUE #1981: Warn on `NaN` loss
Would it not be convenient, particularily for users with less experienced with artificial neural networks, to print a warning when trying to update parameters with a loss of NaN? This could save a lot of wasted resources.

### PR DESCRIPTION
**Pull request by mcabbott**
Closes #1981, by altering `train!` so that when it encounters an infinite / NaN loss, it ~~prints a warning~~ throws an error, and stops. It stops before updating the model, because such an update will usually make everything NaN.

Not sure it should stop, in fact. Maybe it should skip that datapoint & continue?

### PR Checklist

- [x] Tests are added
- [ ] Entry in NEWS.md
- [ ] Documentation, if applicable

### CODE REVIEW

File: test/optimise.jl [hunk 1]
```diff
@@ -87,6 +87,18 @@ end
...
+@testset "Stop on NaN" begin
+  m = Dense(1 => 1)
+  m.weight .= 0
+  CNT = 0
+  Flux.train!(Flux.params(m), 1:100, Descent(0.1)) do i
+    CNT += 1
+    i == 51 ? NaN32 : sum(m([1f0]))
+  end
+  @test CNT == 51  # stopped early
+  @test m.weight[1] ≈ -5  # did not corrupt weights
```
Comment (mcabbott) [inline test/optimise.jl:99]: Somewhat surprisingly, current Flux doesn't actually produce NaN weights, but I don't know where it skips this update:
```
julia> let
       m = Dense(1 => 1)
                m.weight .= 0
                CNT = 0
                Flux.train!(Flux.params(m), 1:100, Descent(0.1)) do i
                  CNT += 1
                  i == 51 ? NaN32 : sum(m([1.0]))
                end
                m.weight, CNT
       end
(Float32[-9.900002;;], 100)
```

Comment (mcabbott) [reply test/optimise.jl:99]: My fault, that iteration just has zero gradient. A more realistic one would be something like this:
```
julia> let
       m = Dense(1 => 1)
                m.weight .= 0
                CNT = 0
                Flux.train!(Flux.params(m), 1:100, Descent(0.1)) do i
                  CNT += 1
                  (i == 51 ? NaN32 : 1f0) * sum(m([1.0]))
                end
                m.weight, CNT
       end
(Float32[NaN;;], 100)
```

**Comment by ToucheSir** [comment]
It would be good to have some way for user code to tell training has stopped prematurely. Not sure what that would look like: return value, callback, flag, etc?

**Comment by mcabbott** [comment]
One way would be to upgrade this to an error. It does depend a little on how complex/comprehensive we think `train!` ought to be.

I found https://github.com/FluxML/Flux.jl/issues/821, in which all options were discussed. And the idea was to write `isfinite(l) || Flux.skip()`, except that this got documented as something to do in a callback, which never worked, it happens too late... and nobody noticed because this was sufficiently confusing? So my opinion is that anything that complicated should be an ordinary `for` loop.

**Comment by ToucheSir** [comment]
Raising `StopException` from the inner loop doesn't seem like the worst idea. Unlike calling `skip` or `stop` in callbacks, this would be able to intercept things in time. I guess the only question is whether making `StopException` part of the public API for this purpose is palatable.

**Comment by mcabbott** [comment]
`StopException` leads directly to `break`, so I think there's nothing gained by implementing an auto-check that way compared to this PR's current code?

The point of the exception was, I now see, was that you could throw it from inside the loss function passed to gradient, and still get to `break`. But documenting it as part of the callback was a mistake.

**Comment by ToucheSir** [comment]
I missed that `StopException` was still being caught by the inner loop of `train!`. In that case, it'd have to be another exception type. Whether adding one just for this makes sense is a good question.

**Comment by mcabbott** [comment]
Ah OK, so you weren't proposing to throw and catch an error. It could just throw a DomainError, that seems the closest Base type.

**Review by ToucheSir — APPROVED** [review]
LGTM, but if @darsnack or @CarloLucibello wouldn't mind having a once-over that would be great.

**Comment by KronosTheLate** [comment]
Thanks for taking the issue I raised so seriously, it is so nice to encounter active developers who listen to user requests. Thanks for your work <3
